### PR TITLE
Update TrinoFileSystemCache to represent latest hadoop implementation

### PR DESCRIPTION
**Pull request by jitheshtr**
## Motivation
Hadoop's implementation of filesystem cache ([hadoop 3.2 filesystem cache](https://github.com/apache/hadoop/blob/release-3.2.0-RC1/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java#L3342)) creates the filesystem object outside of synchronized block. A side effect to this (in addition to reduced locking duration for slow-to-create filesystem implementations) is that there is no interaction between the lock in caching infrastructure and locks internal to a filesystem implementation (during filesystem object creation). We would like to bring this approach to `TrinoFileSystemCache`

A secondary motivation is to improve the concurrency of `TrinoFileSystemCache` operations by avoiding synchronized blocks.

## Description
 - Use ConcurrentHashMap to cache filesystem objects - improves
   concurrency by removing synchronized blocks
 - Filesystem object is created outside cache's lock - similar to latest
   hadoop fs cache impl, further reducing code in critical section.
   Helps with systems where filesystem creation is expensive.
 - Only one thread exclusively creates the filesystem object for a
   given key. Avoids speculative creation and then later discarding of
   filesystem objects compared to hadoop fs cache impl.

There is a more recent update in hadoop 3.3.x branch that limits the number of parallel filesystem object creations using a semaphore. Looking at the description of the issue ([HADOOP-17313](https://issues.apache.org/jira/browse/HADOOP-17313)), it seems to be created as a workaround for speculative-create-and-discard approach used in hadoop implementation which this code avoids.

Benchmark output -
Before:
```
Benchmark                         (numGetCallsPerInvocation)  (numThreads)  (numUsers)  Mode  Cnt   Score   Error  Units
BenchmarkGetFileSystem.benchmark                        1000             1          10  avgt   10   7.747 ± 0.448  ms/op
BenchmarkGetFileSystem.benchmark                        1000             1         100  avgt   10   8.041 ± 0.352  ms/op
BenchmarkGetFileSystem.benchmark                        1000             1        1000  avgt   10   7.492 ± 0.340  ms/op
BenchmarkGetFileSystem.benchmark                        1000            16          10  avgt   10  69.900 ± 8.675  ms/op
BenchmarkGetFileSystem.benchmark                        1000            16         100  avgt   10  66.847 ± 2.937  ms/op
BenchmarkGetFileSystem.benchmark                        1000            16        1000  avgt   10  70.222 ± 4.286  ms/op
```
After:
```
Benchmark                         (numGetCallsPerInvocation)  (numThreads)  (numUsers)  Mode  Cnt   Score   Error  Units
BenchmarkGetFileSystem.benchmark                        1000             1          10  avgt   10   7.767 ± 0.511  ms/op
BenchmarkGetFileSystem.benchmark                        1000             1         100  avgt   10   7.412 ± 0.267  ms/op
BenchmarkGetFileSystem.benchmark                        1000             1        1000  avgt   10   7.385 ± 0.284  ms/op
BenchmarkGetFileSystem.benchmark                        1000            16          10  avgt   10  26.333 ± 2.696  ms/op
BenchmarkGetFileSystem.benchmark                        1000            16         100  avgt   10  28.163 ± 1.489  ms/op
BenchmarkGetFileSystem.benchmark                        1000            16        1000  avgt   10  29.545 ± 4.080  ms/op
```
(above results are from an 8 core intel macbook pro)

> Is this change a fix, improvement, new feature, refactoring, or other?

Improvement

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Update the implementation of TrinoFileSystemCache class in Hive connector.

> How would you describe this change to a non-technical end user or system administrator?

Bring `TrinoFileSystemCache` implementation inline with latest hadoop implementation and improve cache performance in the process.

## Documentation

( ) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Hive
* Improve performance by reducing contention in Trino's file system cache. ({issue}`13243 `)
```

### CODE REVIEW

**Comment by jitheshtr** [comment]
Todo -
1. Rebase with master
2. Fix CI failures due to interaction between various product tests and the new test added here, both accessing global TrinoFileSystemCache within the same JVM

File: lib/trino-hdfs/src/main/java/io/trino/hdfs/TrinoFileSystemCache.java [hunk 1]
```diff
@@ -102,55 +96,45 @@ public FileSystem getUnique(URI uri, Configuration conf)
...
+                    if (currFileSystemHolder.credentialsChanged(uri, conf, privateCredentials)) {
+                        return new FileSystemHolder(uri, conf, privateCredentials);
+                    }
+                    else {
+                        return currFileSystemHolder;
+                    }
+                }
+            });
+
+            fileSystemHolder.createFileSystemOnce();
```
Comment (phd3) [inline lib/trino-hdfs/src/main/java/io/trino/hdfs/TrinoFileSystemCache.java:131]: Can we add a comment here why this is outside of the cache compute? Seems like that's an important piece of why this works

**Review by phd3 — COMMENTED** [review]
some initial comments

File: lib/trino-hdfs/src/main/java/io/trino/hdfs/TrinoFileSystemCache.java [hunk 2]
```diff
@@ -102,55 +96,45 @@ public FileSystem getUnique(URI uri, Configuration conf)
     @VisibleForTesting
     int getCacheSize()
     {
-        return map.size();
+        return cache.size();
```
[the remark this answers is not in the document]
Comment (jitheshtr) [reply lib/trino-hdfs/src/main/java/io/trino/hdfs/TrinoFileSystemCache.java:99]: It is possible that `cacheSize` and `cache.size()` values can differ momentarily during file system cache operations, for eg: if a thread A invokes `getCacheSize()` while another thread B is just after `cacheSize.getAndUpdate()` (inside `getInternal()`) and before the key is added to`cache` datastructure - the `cacheSize` and `cache.size()` values observed by thread A at this moment will be different. Similar discrepancy can happen when multiple `getInternal()` and `closeFileSystem()`s invocations are running concurrently. Eventually, `cacheSize` and `cache.size()` becomes equal.

It is true that `cacheSize` is the real tracker for the number of elements in the cache, but `cache.size()` was used here as it is the real data structure holding the keys, and reflects the "real" cache entry count. Either would work for this use case - knowing the discrepancy between these two mentioned above.

File: lib/trino-hdfs/src/main/java/io/trino/hdfs/TrinoFileSystemCache.java [hunk 3]
```diff
@@ -102,55 +96,45 @@ public FileSystem getUnique(URI uri, Configuration conf)
...
-            catch (IOException e) {
-                stats.newGetCallFailed();
-                throw e;
-            }
+        int maxSize = conf.getInt("fs.cache.max-size", 1000);
+        FileSystemHolder fileSystemHolder;
+        try {
+            fileSystemHolder = cache.compute(key, (k, currFileSystemHolder) -> {
+                if (currFileSystemHolder == null) {
+                    if (cacheSize.getAndUpdate(curr -> curr < maxSize ? (curr + 1) : curr) >= maxSize) {
```
[the remark this answers is not in the document]
Comment (jitheshtr) [reply lib/trino-hdfs/src/main/java/io/trino/hdfs/TrinoFileSystemCache.java:114]: We use `>=` as `getAndUpdate()` returns the previous value -
```
Atomically updates the current value with the results of applying the given function, returning the previous value.
```

This is captured in the test case `testFileSystemCacheException()`

File: lib/trino-hdfs/src/main/java/io/trino/hdfs/TrinoFileSystemCache.java [hunk 4]
```diff
@@ -306,23 +313,49 @@ public String toString()
 
     private static class FileSystemHolder
     {
-        private final FileSystem fileSystem;
+        private final URI uri;
```
[the remark this answers is not in the document]
Comment (jitheshtr) [reply lib/trino-hdfs/src/main/java/io/trino/hdfs/TrinoFileSystemCache.java:316]: The thought process was to keep uri and conf provided by the thread who created the FileSystemHolder key - which is what happens in existing implementation. `createFileSystemOnce()` could be called by a different thread having a different uri and/or conf object due to the original thread being scheduled out of execution by operating system just before invoking `createFileSystemOnce()`.

File: lib/trino-hdfs/src/test/java/io/trino/hdfs/TestFileSystemCache.java [hunk 5]
```diff
@@ -56,9 +77,110 @@ public void testFileSystemCache()
...
+        private final FileSystemConsumer consumer;
+
+        private HdfsEnvironment environment = new HdfsEnvironment(
+                new DynamicHdfsConfiguration(new HdfsConfigurationInitializer(new HdfsConfig()), ImmutableSet.of()),
+                new HdfsConfig(),
+                new ImpersonatingHdfsAuthentication(new SimpleHadoopAuthentication(), new SimpleUserNameProvider()));
+
+        CreateFileSystemAndConsume(SplittableRandom random, int numUsers, int numGetCallsPerInvocation, FileSystemConsumer consumer)
+        {
+            this.random = random;
```
Comment (phd3) [inline lib/trino-hdfs/src/test/java/io/trino/hdfs/TestFileSystemCache.java:170]: nit: null check since this is also exposed outside of the class

**Review by phd3 — COMMENTED** [review]
The implementation looks good to me, but IMO would be useful to get some more 👀 on this as well since the change is a bit involved.

**Comment by phd3** [comment]
cc @electrum

File: lib/trino-hdfs/src/main/java/io/trino/hdfs/TrinoFileSystemCache.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jitheshtr) [reply lib/trino-hdfs/src/main/java/io/trino/hdfs/TrinoFileSystemCache.java:323]: Yes, should be fine, will update - we can avoid storing uri/conf in `FileSystemHolder` with this change.

File: lib/trino-hdfs/src/test/java/io/trino/hdfs/TestFileSystemCache.java [hunk 6]
```diff
@@ -56,9 +77,110 @@ public void testFileSystemCache()
...
         return environment.getFileSystem(identity, new Path("/"), newEmptyConfiguration());
     }
+
+    @FunctionalInterface
+    public interface FileSystemConsumer
+    {
+        void consume(FileSystem fileSystem) throws IOException;
+    }
+
+    private static class FileSystemCloser
```
[the remark this answers is not in the document]
Comment (jitheshtr) [reply lib/trino-hdfs/src/test/java/io/trino/hdfs/TestFileSystemCache.java:143]: If we use a lambda, we get the error `Hadoop FileSystem instances are shared and should not be closed`. Had to add `FileSystemCloser` and annotate its `consume()` method with `@SuppressModernizer` to fix this.

**Comment by sopel39** [comment]
please rebase

File: lib/trino-hdfs/src/test/java/io/trino/hdfs/BenchmarkGetFileSystem.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+@Fork(2)
+@Measurement(iterations = 5, time = 5000, timeUnit = MILLISECONDS)
+@BenchmarkMode(Mode.AverageTime)
+public class BenchmarkGetFileSystem
+{
+    @State(Scope.Thread)
+    public static class BenchmarkData
+    {
+        @Param({"10", "100", "1000"})
+        private int numUsers;
```
Comment (phd3) [inline lib/trino-hdfs/src/test/java/io/trino/hdfs/BenchmarkGetFileSystem.java:56]: nit: `userCount`, `threadCount`, `getCallsPerInvocation` to avoid abbreviations;

File: lib/trino-hdfs/src/test/java/io/trino/hdfs/BenchmarkGetFileSystem.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+            this.executor = Executors.newFixedThreadPool(numThreads);
+        }
+
+        @TearDown(Level.Invocation)
+        public void tearDown()
+                throws IOException
+        {
+            TrinoFileSystemCache.INSTANCE.closeAll();
+            executor.shutdown();
```
Comment (phd3) [inline lib/trino-hdfs/src/test/java/io/trino/hdfs/BenchmarkGetFileSystem.java:87]: `shutdownNow()` as we do not need to wait here

File: lib/trino-hdfs/src/test/java/io/trino/hdfs/BenchmarkGetFileSystem.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+        }
+    }
+
+    @Benchmark
+    public void benchmark(BenchmarkData data)
+            throws InterruptedException, ExecutionException
+    {
+        List<Future<Void>> futures = data.executor.invokeAll(data.callableTasks);
+        for (Future<Void> fut : futures) {
```
Comment (phd3) [inline lib/trino-hdfs/src/test/java/io/trino/hdfs/BenchmarkGetFileSystem.java:96]: this can be simplified (simile to other comment)

File: lib/trino-hdfs/src/test/java/io/trino/hdfs/TestFSDataInputStreamTail.java [hunk 10]
```diff
@@ -70,6 +70,7 @@ public void tearDown()
         closeAll(
                 () -> fs.delete(new Path(tempRoot.toURI()), true),
                 fs);
+        fs = null;
```
Comment (phd3) [inline lib/trino-hdfs/src/test/java/io/trino/hdfs/TestFSDataInputStreamTail.java:73]: why is this a related change ? Can we keep this in a separate commit ?

File: lib/trino-hdfs/src/test/java/io/trino/hdfs/TestFileSystemCache.java [hunk 11]
```diff
@@ -56,9 +79,101 @@ public void testFileSystemCache()
...
+                            1000,
+                            new FileSystemCloser()));
+        }
+        ExecutorService executor = Executors.newFixedThreadPool(numThreads);
+
+        assertEquals(TrinoFileSystemCache.INSTANCE.getFileSystemCacheStats().getCacheSize(), 0);
+        List<Future<Void>> futures = executor.invokeAll(callableTasks);
+        for (Future<Void> fut : futures) {
+            fut.get();
+        }
```
Comment (phd3) [inline lib/trino-hdfs/src/test/java/io/trino/hdfs/TestFileSystemCache.java:120]: executor.invokeAll(callableTasks).forEach(MoreFutures::getFutureValue);

File: lib/trino-hdfs/src/test/java/io/trino/hdfs/TestFileSystemCache.java [hunk 12]
```diff
@@ -56,9 +79,101 @@ public void testFileSystemCache()
...
+
+    public static class CreateFileSystemsAndConsume
+            implements Callable<Void>
+    {
+        private final SplittableRandom random;
+        private final int numUsers;
+        private final int numGetCallsPerInvocation;
+        private final FileSystemConsumer consumer;
+
+        private HdfsEnvironment environment = new HdfsEnvironment(
```
Comment (phd3) [inline lib/trino-hdfs/src/test/java/io/trino/hdfs/TestFileSystemCache.java:156]: private static final

File: lib/trino-hdfs/src/test/java/io/trino/hdfs/TestFileSystemCache.java [hunk 13]
```diff
@@ -56,9 +79,101 @@ public void testFileSystemCache()
...
+                new DynamicHdfsConfiguration(new HdfsConfigurationInitializer(new HdfsConfig()), ImmutableSet.of()),
+                new HdfsConfig(),
+                new ImpersonatingHdfsAuthentication(new SimpleHadoopAuthentication(), new SimpleUserNameProvider()));
+
+        CreateFileSystemsAndConsume(SplittableRandom random, int numUsers, int numGetCallsPerInvocation, FileSystemConsumer consumer)
+        {
+            this.random = requireNonNull(random, "random is null");
+            this.numUsers = numUsers;
+            this.numGetCallsPerInvocation = numGetCallsPerInvocation;
```
Comment (phd3) [inline lib/trino-hdfs/src/test/java/io/trino/hdfs/TestFileSystemCache.java:165]: same comment about naming

**Review by phd3 — APPROVED** [review]
Final set of comments - looks good to me

File: lib/trino-hdfs/src/test/java/io/trino/hdfs/TestFSDataInputStreamTail.java (hunk 10 above)
Comment (jitheshtr) [reply lib/trino-hdfs/src/test/java/io/trino/hdfs/TestFSDataInputStreamTail.java:73]: This came in as part of bringing `trino-testing-services` dependency into `trino-hdfs`. Resource deallocation check added via `ManageTestResources` (in PR #15165) got activated in `lib/trino-hdfs` resulting in this test failure here. Adding `fs = null;` fixes this, as was the case with similar changes in the PR referenced. Can move this to a different commit

**Review by phd3 — APPROVED** [review]
